polo-common/src/main/java/dev/dhdf/polo/util/Sync.java

```diff
@@ -11,13 +14,36 @@
  */
 public class Sync implements Runnable {
     private final WebClient client;
+    private final Logger logger = LoggerFactory.getLogger(Sync.class);
+
+    private boolean get_events;
 
     public Sync(WebClient client) {
         this.client = client;
+        this.get_events = true;
     }
 
     @Override
     public void run() {
-        this.client.getChat();
+        boolean success = false;
+
+        // Default to the events endpoint unless we know it isn't supported
+        if (get_events)
+            success = this.client.getEvents();
+
+        // Fall back to the legacy plain chat endpoint
+        if (!success) {
+            success = this.client.getChat();
+
+            if (!success) {
+                // Even the legacy endpoint failed, next time try the events
+                // endpoint too
+                get_events = true;
+            } else  if (get_events) {
+                // It worked! just use this in future
+                logger.warn("Falling back to legacy chat endpoint, please update matrix-appservice-minecraft");
+                get_events = false;
+            }
+        }
```

Comment:
This could be simplified:

```suggestion
        // Default to the events endpoint unless we know it isn't supported
        if (get_events && !this.client.getEvents() && this.client.getChat()) {
            // Events failed, but legacy succeeded: use legacy endpoint in the future
            logger.warn("Falling back to legacy chat endpoint, please update matrix-appservice-minecraft");
            get_events = false;
        } else {
            this.client.getChat();
        }
```

Also, the method names are misleading since they actually process the events and chat respectively, but IMO that's for a different PR.

Reply:
Yes, this is convoluted from when the logic was in getChat().

well, if we lose connection to the appservice, it might be getChat() that fails even when get_events is true, then the else block would try a 2nd time, probably not what we want. Also I wanted it to be able to try the events API again after both failed, e.g. if the user went and upgraded matrix-appservice-minecraft.

I think a strict logical simplification would be:
``` java
        // Default to the events endpoint unless we know it isn't supported
        if (!supportsEvents || !this.client.getEvents()) {
            // Fall back to the legacy plain chat endpoint
            if (!this.client.getChat()) {
                // Even the legacy endpoint failed, next time try the events
                // endpoint too
                supportsEvents = true;
            } else if (supportsEvents) {
                // It worked! just use this in future
                logger.warn("Falling back to legacy chat endpoint, please update matrix-appservice-minecraft");
                supportsEvents = false;
            }
        }
```

Reply:
Random thought: instead of detecting it here, perhaps implement versioning in the protocol?

Reply:
My first version just changed /get/. arguably its not a stable api so could we just get away with replacing it?

What sort of versioning were you thinking? the appservice reporting what versions it supports and the plugin using the preferred apis it knows are available? maybe if something goes wrong or it can't connect it does another version query.

Reply:
> My first version just changed /get/. arguably its not a stable api so could we just get away with replacing it?

Probably could, yeah.

> What sort of versioning were you thinking? the appservice reporting what versions it supports and the plugin using the preferred apis it knows are available? maybe if something goes wrong or it can't connect it does another version query.

I was thinking either something similar to what Matrix does, or returning a semantic version with `/vibeCheck` to check against.

Reply:
My preference at the moment is to leave as is (the simplification above). it works well enough IMO. Any strong objections?